mpsl: cx: Add a generic coex variant using a single GPIO (#16716)
nrfconnect/sdk-nrf: doc/nrf/device_guides/wifi_coex.rst

   * :kconfig:option:`CONFIG_MPSL_CX`
   * :kconfig:option:`CONFIG_MPSL_CX_3WIRE`

.. _ug_radio_mpsl_cx_generic_1wire:

Generic one wire coexistence
============================

Refer to :ref:`ug_radio_coex_mpsl_cx_based` for the general requirements of this implementation.

> The generic one wire coexistence interface can be used to allow the :ref:`SoftDevice Controller <nrfxlib:softdevice_controller>` to coexist alongside an LTE device on a separate chip.

Reviewer: Not sure we should mention SoftDevice Controller explicitly here as this interface should work fine for 802.15.4 as well.
Right @ankuns?

Reviewer 2: Generally it should.

Author: Shall we just mention both SDC and 802.15.4 ? This line just gives example use-cases

> It is specifically designed for the coex interface of the nRF91 Series SiP.

Reviewer: I'm a bit confused by the title being generic one wire interface, but this documentation saying it is for LTE devices and is specifically designed for the coex interface of the nRF91 series.

Could it be an idea to split it as is done for the generic 3 wire and nrf700x coex interface?

Author: We can rephrase for instance to say that it's "compatible with the coex interface for the nrf91". I just wanted to make sure customers aren't confused when we deprecate the other one

   ============  =========  =================================  ==============

Enabling generic one wire coexistence
-------------------------------------

To enable the generic one wire coexistence, do the following:

1. Add the following node to the devicetree source file:

   .. code-block::

      / {
            nrf_radio_coex: radio_coex_one_wire {
               status = "okay";
               compatible = "generic-radio-coex-one-wire";
               grant-gpios =   <&gpio0 25 GPIO_ACTIVE_LOW>;
         };
      };

#. Optionally, if not using the nRF91 Series SiP, the ``GRANT`` signal may be configured active-high using ``GPIO_ACTIVE_HIGH``
#. Optionally, replace the node name ``radio_coex_one_wire`` with a custom one.
#. Replace the pin number provided for the ``grant-gpios`` property:
   This is the GPIO characteristic of the device that interfaces with the ``GRANT`` signal of the PTA (RF medium access granted).

   The first element ``&gpio0`` indicates the GPIO port (``port 0`` has been selected in the example shown).
   The second element is the pin number on that port.

#. On the nRF5340, apply the same devicetree node mentioned in Step 1 to the network core.

>    Apply the overlay to the correct network-core child image by creating an overlay file named :file:`child_image/*childImageName*.overlay` in your application directory, for example :file:`child_image/multiprotocol_rpmsg.overlay`.

Reviewer: Discussed offline. child_image shouldn't be used any more. We need update the documentation to use sysbuild and figure out the state of multiprotocol_rpmsg and 802154_rpmsg as I'm not sure those has been updated lately.

Author: After reading through other docs a bit, I wasn't sure what to write. So I ended up removing these docs. We can always add sysbuild instructions later (and clean up the information about child images from the other docs)

.. table:: Generic one wire coexistence protocol pins

   ============  =========  =================================  ==============
   Pin           Direction  Description                        DT property
   ============  =========  =================================  ==============
   GRANT         In         Grant signal                       grant-gpios
   ============  =========  =================================  ==============

Enabling generic one wire coexistence
-------------------------------------

To enable the generic one wire coexistence, do the following:

1. Add the following node to the devicetree source file:

   .. code-block::

      / {
            nrf_radio_coex: radio_coex_one_wire {
               status = "okay";
               compatible = "generic-radio-coex-one-wire";
               grant-gpios =   <&gpio0 25 GPIO_ACTIVE_LOW>;
         };
      };

#. Optionally, if not using the nRF91 Series SiP, the ``GRANT`` signal may be configured active-high using ``GPIO_ACTIVE_HIGH``
#. Optionally, replace the node name ``radio_coex_one_wire`` with a custom one.

> #. Replace the pin numbers provided for each of the required properties:
> 
>    * ``grant-gpios`` - GPIO characteristic of the device that controls the ``GRANT`` signal of the PTA (RF medium access granted).

Reviewer 3: ```suggestion
#. Replace the pin number provided for the ``grant-gpios`` property.
   This is the GPIO characteristic of the device that controls the ``GRANT`` signal of the PTA (RF medium access granted).
```
Since there's only one property to configure.

Author: also replaced "controls" with "interfaces with" because it's an input GPIO
